mistakedetection.tests/src/ca/mcgill/sel/mistakedetection/tests/MistakeDetectionPerformanceAnalysis.java

```diff
@@ -36,387 +67,235 @@
+         "G15_3", "G15_7", "G15_16", "G16_9", "G16_11", "G16_12", "G17_5", "G17_8", "G17_12", "G17_25",
+         "G12_3", "G12_14", "G13_5", "G13_9", "G13_15", "G14_2", "G14_7", "G14_12", "G15_5", "G15_9");
+    // list of solutions that should be (re)evaluated by (re)generating a spreadsheet for them
+    List<String> idsToEvaluate = List.of("G15_9"); // or List.copyOf(solutionIds)
+    // solutions not found on the local filesystem, which can be ignored
+    List<String> notFoundIds = new ArrayList<>();
+    // existing solutions that caused the MDS to crash
+    List<String> failedIds = new ArrayList<>();
+
+    solutionIds.stream().filter(Predicate.not(String::isEmpty)).forEach(solutionId -> {
+      try {
+        var studentClassDiagram = getHotelStudentClassDiagram(solutionId);
+        if (studentClassDiagram == null) {
+          notFoundIds.add(solutionId);
+        } else {
+          var studentSolution = studentSolutionFromClassDiagram(studentClassDiagram);
+          var comparison = MistakeDetection.compare(instructorSolution, studentSolution, true);
+          assertNotNull(comparison);
+          if (idsToEvaluate.contains(solutionId)) {
+            produceExcelSheet(comparison, solutionId, HOTEL_OUTPUT_SPREADSHEET_LOC);
+          }
+        }
+      } catch (Exception e) {
+        System.err.println("Could not detect mistakes for solution " + solutionId + " due to error:");
+        e.printStackTrace();
+        failedIds.add(solutionId);
+      }
+    });
+    if (!notFoundIds.isEmpty()) {
+      System.err.println("Could not find cdm files for " + notFoundIds.size() + " solutions: " + notFoundIds);
+    }
+    if (!failedIds.isEmpty()) {
+      fail("Could not detect mistakes for " + failedIds.size() + " solutions: " + failedIds);
+    }
   }
 
-  @Test
-  public void testStudentSolution28() throws Exception {
-    var name = "G14_12";
-    var studentClassDiagram = getStudentClassDiagram(name);
-    var studentSolution = studentSolutionFromClassDiagram(studentClassDiagram);
-
-    var comparison = MistakeDetection.compare(instructorSolution, studentSolution, true);
-
+  /** Tests that the MDS runs without errors on the submissions from the final exam dataset. */
+  @Test public void testThatMdsRunsOnFinalExamStudentSolutions() {
+    var ma = maf.createModelingAssistant();
+    Map<Integer, List<Integer>> solutionsToNumMistakes = new TreeMap<>();
+    var instSolutions = setupSmartHomeInstructorSolutions();
+    for (var instSolution: instSolutions) {
+      try (var files = Files.walk(Path.of(FINAL_EXAM_SUBMISSIONS_PATH), 2)) {
+        List<String> validCdms = new ArrayList<>();
+        List<String> invalidCdms = new ArrayList<>();
+        var filenamePattern = Pattern.compile("\\d+\\.cdm");
+        var studentCdms = files
+            .filter(file ->
+                !Files.isDirectory(file) && filenamePattern.matcher(file.getFileName().toString()).matches()
+                && !file.endsWith("0.cdm"))
```

Comment:
In Java, `File("dir/0.cdm").endsWith("0.cdm")` is true but `File("dir/20.cdm").endsWith("0.cdm")` is false, unlike with strings